design/control-data-plane-separation/design.md

```diff
@@ -23,377 +23,118 @@ architecture does not allow the control plane and data plane to scale independen
-    f5.nginx.agent.sdk.events.EventReport event_report = 12 [(gogoproto.jsontag) = "event_report"];
+#### Upstream server state
 
-    // DataplaneSoftwareDetails contains details for additional software running on the dataplane that pertains
-    // to NGINX Agent
-    DataplaneSoftwareDetails dataplane_software_details = 13 [(gogoproto.jsontag) = "dataplane_software_details"];
+In the current implementation using NGINX Plus, when only upstream servers change, NGF writes the upstream servers in the nginx config, but does not reload. It then calls the NGINX Plus API to update the servers. This process allows us to update the upstream servers using the API and not have to reload nginx, while still having the upstream servers exist in the nginx config for easy debugging and consistency. However, when using agent, any config write will result in a reload. To preserve the ability to update upstreams with the API without needing a reload, we'll have to utilize a `state` file instead of writing the servers directly in the nginx config. This way the list of servers is still available on the filesystem for debugging, but is written by the nginx when making the API call instead of by the control plane directly.
 
-    // DataplaneUpdate contains details for dataplane resources that have changed
-    DataplaneUpdate dataplane_update = 14 [(gogoproto.jsontag) = "dataplane_update"];
-  }
-}
-```
-
-There are many command types defined above, and the control plane does not care about all of them. However, any commands
-that we do not care about can be easily ignored by the control plane by inspecting the message type.
+An example of what this looks like is defined [here](https://docs.nginx.com/nginx/admin-guide/load-balancer/dynamic-configuration-api/#configuring-persistence-of-dynamic-configuration).
 
-For the full command message definition, see
-this [file](https://github.com/nginx/agent/blob/main/sdk/proto/command.proto).
+#### Secret duplication and synchronization
 
-#### Nginx Config Message
+There are multiple Secrets that an NGINX Plus user can and will be creating. These include:
 
-The control plane sends the `NginxConfig` message in chunks over the `Download` channel to the agent in order to
-configure nginx.
-
-```protobuf
-message NginxConfig {
-  NginxConfigAction action = 1 [(gogoproto.jsontag) = "action"];
-  ConfigDescriptor config_data = 2 [(gogoproto.jsontag) = "config_data"];
-  ZippedFile zconfig = 3 [(gogoproto.jsontag) = "zconfig"];
-  ZippedFile zaux = 4 [(gogoproto.jsontag) = "zaux"];
-  AccessLogs access_logs = 5 [(gogoproto.jsontag) = "access_logs"];
-  ErrorLogs error_logs = 6 [(gogoproto.jsontag) = "error_logs"];
-  SslCertificates ssl = 7 [(gogoproto.jsontag) = "ssl"];
-  DirectoryMap directory_map = 8 [(gogoproto.jsontag) = "directory_map"];
-}
-```
+- JWT Secret for running NGINX Plus
+- Docker Registry Secret for pulling NGINX Plus
+- Client cert/key Secret for NIM connection
+- CA cert Secret for NIM connection
 
```

Comment:
What about a Secret for NGF -> Agent connection?

Reply:
Yeah, I'm wavering on this one as to whether a user should create it it before they create their Gateway, in that namespace, or if they create it once and we duplicate it. If it's managed by something like cert-manager, trying to think what the best option would be.

This would apply for all users, not just NGINX Plus.